src/internal/_equals.js

```diff
@@ -14,10 +14,19 @@ module.exports = function _equals(a, b, stackA, stackB) {
     return false;
   }
 
+  if (a == null || b == null) {
+    return false;
+  }
+
+  if (typeof a.equals === 'function') {
+    return a.equals(b);
+  }
+
+  if (typeof b.equals === 'function') {
+    return b.equals(a);
+  }
+
```

Comment:
Where does responsibility lie for ensuring that both values have the same constructor? In Ramda's `_equals` function, or in the various `equals` methods to which it dispatches?

This is another example of the issue @CrossEye raised above:

```javascript
function Point(x, y) {
  this.x = x;
  this.y = y;
}
Point.prototype.equals = function(point) {
  return this instanceof Point &&
         this.x === point.x && this.y === point.y;
};

ColorPoint = function(x, y, color) {
  this.x = x;
  this.y = y;
  this.color = color;
}
ColorPoint.prototype = new Point(0, 0);
ColorPoint.prototype.equals = function(point) {
  return point instanceof ColorPoint &&
         this.x === point.x && this.y === point.y &&
         this.color === point.color;
};

var p = new Point(2, 2);
var cp = new ColorPoint(2, 2, 'red');

p.equals(cp);  // => true
cp.equals(p);  // => false
```

The question is whether in the OO world it makes sense for `p.equals(cp)` to evaluate `true`, or whether `this instanceof Point` should be replaced with `this.constructor === Point` to ensure that a ColorPoint instance is never considered equal to a Point instance.

Regardless of the answer to this question I believe we should ensure that both values have the same constructor in `_equals`. `x` ≍ `y` implies `x.constructor` ≍ `y.constructor` (for every "accessible" value of `x`). If I have a function which returns `x.constructor.name` and you tell me that `x` and `y` are equivalent, I should be able to replace `x` with `y` without any observable effect. If I cannot, `x` and `y` are clearly not equivalent.

One option is to add this after the `a == null || b == null` check:

```javascript
if (a.constructor !== b.constructor) {
  return false;
}
```

We could then be sure that `a` and `b` are of the same type, so any (non-broken) `equals` method would guarantee `a.equals(b) === b.equals(a)`.

Unfortunately, this change would prevent values created in different environments from being considered equal, which is undesirable (`[1, 2, 3]` in an iframe should be equal to `[1, 2, 3]` in the parent frame). We could instead compare `constructor.name`, but there's no guarantee that two constructors with the same name are equivalent.

I don't have a solution to the cross-frame issue, so @buzzdecafe's suggestion of checking both `a.equals(b)` and `b.equals(a)` seems the best course of action.